storage/BloomFilterIndexSubBlock.hpp

```diff
@@ -0,0 +1,184 @@
+  *        block memory amongst sub-blocks.
+  * @warning description must be valid. DescriptionIsValid() should be called
+  *          first if necessary.
+  *
+  * @param relation The relation tuples belong to.
+  * @param description A description of the parameters for this type of
+  *        IndexSubBlock.
+  * @return The average/ammortized number of bytes used to index a single
+  *         tuple of relation in an IndexSubBlock of this type described by
+  *         description.
+  **/
+  static std::size_t EstimateBytesPerTuple(const CatalogRelationSchema &relation,
+                                          const IndexSubBlockDescription &description);
+
+  /**
+   * @brief Estimate the total number of bytes (including any applicable
+   *        overhead) occupied by this IndexSubBlock within the StorageBlock.
+   *        This function is to be used by those indicies whose occupied size
+   *        in the block does not depend on the number of tuples being indexed.
+   * @warning description must be valid. DescriptionIsValid() should be called
+   *          first if necessary.
+   * @note This function will be invoked by StorageBlockLayout::finalize()
+   *       if and only when EstimateBytesPerTuple() returns a zero size.
+   *
+   * @param relation The relation tuples belong to.
+   * @param description A description of the parameters for this type of
+   *        IndexSubBlock.
+   * @return The total number of bytes occupied by this IndexSubBlock within
+   *         the StorageBlock.
+   **/
+  static std::size_t EstimateBytesPerBlock(const CatalogRelationSchema &relation,
+                                           const IndexSubBlockDescription &description);
+
+  IndexSubBlockType getIndexSubBlockType() const override {
+    return kBloomFilter;
+  }
+
+  bool supportsAdHocAdd() const override {
+    return false;
+  }
+
+  bool supportsAdHocRemove() const override {
+    return false;
+  }
+
+  bool addEntry(const tuple_id tuple) override;
+
+  bool bulkAddEntries(const TupleIdSequence &tuples) override;
+
+  void removeEntry(const tuple_id tuple) override;
+
+  void bulkRemoveEntries(const TupleIdSequence &tuples) override;
+
+  predicate_cost_t estimatePredicateEvaluationCost(
+     const ComparisonPredicate &predicate) const override;
+
+  /**
+  * @note Currently this version only supports predicates with
+  *       equality comparison of a static literal value with a non-composite key.
+  **/
+  TupleIdSequence* getMatchesForPredicate(const ComparisonPredicate &predicate,
+                                          const TupleIdSequence *filter) const override;
+
+  bool rebuild() override;
+
+  /**
+   * @brief Get the selectivity provided by the index for a given predicate.
+   *        There are only two selectivity levels returned by this function-
+   *        either a zero(none) selectivity for bloom filter miss
+   *        or a one(all) selectivity for bloom filter hit.
```

Comment:
Might want to add that a 'hit' means that the predicate possibly returns tuples.